infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironmentFactory.java

```diff
@@ -149,6 +160,45 @@ protected KubernetesEnvironment doCreate(
     return k8sEnv;
   }
 
+  /**
+   * Merges the specified pods and deployments to a single Deployment.
+   *
+   * <p>Note that method will modify the specified collections and put work result there.
+   *
+   * @param pods pods to merge
+   * @param deployments deployments to merge
+   * @param services services to reconfigure to point new deployment
+   * @throws ValidationException if the specified lists has pods with conflicting configuration
+   */
+  private void mergePods(
+      Map<String, Pod> pods, Map<String, Deployment> deployments, Map<String, Service> services)
+      throws ValidationException {
+    List<PodData> podsData =
+        Stream.concat(
+                pods.values().stream().map(PodData::new),
+                deployments.values().stream().map(PodData::new))
+            .collect(Collectors.toList());
+
+    Deployment deployment = podMerger.merge(podsData);
+    String deploymentName = deployment.getMetadata().getName();
+
+    // provision merged deployment instead of recipe pods/deployments
+    pods.clear();
+    deployments.clear();
+    deployments.put(deploymentName, deployment);
+
+    // multiple pods/deployments are merged to one deployment
+    // to avoid issues because of overriding labels
+    // provision const label and selector to match all services to merged Deployment
+    deployment
+        .getSpec()
+        .getTemplate()
+        .getMetadata()
+        .getLabels()
+        .put(DEPLOYMENT_NAME_LABEL, deploymentName);
+    services.values().forEach(s -> setSelector(s, DEPLOYMENT_NAME_LABEL, deploymentName));
```

Comment:
Question: Since we're resetting the service's selector, is it possible now for a collision to happen if two pods expose the same port?

e.g.
```
Pod1: label:pod1=pod1, port: 8080
Pod2: label:pod2=pod2, port: 8080
Service1: selector: pod1=pod1, port: 8080
[...]
```
E.g. if in your sample devfile, if both pod/deployments exposed port `3000`, would it mean that the service load balances requests between nodejs and mongo unintentionally?

Reply:
Yes, because all containers in a pod share the same network stack and therefore also port assignments.

Reply:
This also means that it is impossible to have a clustered server of some sort as part of the workspace.

Reply:
> would it mean that the service load balances requests between nodejs and mongo unintentionally?

**NO**. As Lukas mentioned, all containers in one pods shared the same network, and if two processes in the pod (even in different containers) will try to use the same port, then for one of them `port is already is use` error will happen.

Reply:
But I mean pods, not containers -- can we not have a deployment that contains two pods that expose the same port?

Reply:
As far as I know, Deployment can have only ONE pod

Reply:
Right, of course.